internal/pr/promote/promotion.go

```diff
@@ -0,0 +1,99 @@
+package promote
+
+import (
+	"fmt"
+	"github.com/nestoca/joy/api/v1alpha1"
+)
+
+type Promotion struct {
+	// GitBranchProvider is the provider for managing git branches.
+	BranchProvider BranchProvider
+
+	// PullRequestProvider is the provider of pull requests.
+	PullRequestProvider PullRequestProvider
+
+	// Prompt is the prompt to use for user interaction.
+	Prompt Prompt
+}
+
+func NewPromotion(branchProvider BranchProvider, pullRequestProvider PullRequestProvider, prompt Prompt) *Promotion {
+	return &Promotion{
+		BranchProvider:      branchProvider,
+		PullRequestProvider: pullRequestProvider,
+		Prompt:              prompt,
+	}
+}
+
+func NewDefaultPromotion() *Promotion {
+	return NewPromotion(
+		&GitBranchProvider{},
+		&GitHubPullRequestProvider{},
+		&SurveyPrompt{},
+	)
+}
+
+// Promote prompts user to create a pull request for current branch and to select environment to auto-promote builds
+// of pull request to, and then configures the pull request accordingly.
+func (p *Promotion) Promote(environments []*v1alpha1.Environment) error {
+	branch, err := p.BranchProvider.GetCurrentBranch()
+	if err != nil {
+		return fmt.Errorf("getting current branch: %w", err)
+	}
+
+	if branch == "master" || branch == "main" {
+		p.Prompt.PrintMasterBranchPromotion()
+		return nil
+	}
+
+	exists, err := p.PullRequestProvider.Exists(branch)
+	if err != nil {
+		return fmt.Errorf("checking if pull request exists for branch %s: %w", branch, err)
+	}
+
+	if !exists {
+		shouldCreate, err := p.Prompt.WhetherToCreateMissingPullRequest()
+		if err != nil {
+			return fmt.Errorf("prompting user to create pull request: %w", err)
+		}
+		if !shouldCreate {
+			p.Prompt.PrintNotCreatingPullRequest()
+			return nil
+		}
+
+		err = p.PullRequestProvider.CreateInteractively(branch)
+		if err != nil {
+			return fmt.Errorf("creating pull request: %w", err)
+		}
+	}
+
+	env, err := p.PullRequestProvider.GetPromotionEnvironment(branch)
+	if err != nil {
+		return fmt.Errorf("getting promotion environment for branch %s pull request: %w", branch, err)
+	}
+
+	promotableEnvironmentNames := getPromotableEnvironmentNames(environments)
+	env, err = p.Prompt.WhichEnvironmentToPromoteTo(promotableEnvironmentNames, env)
+	if err != nil {
+		return fmt.Errorf("prompting user to select promotion environment: %w", err)
+	}
+
+	if err := p.PullRequestProvider.SetPromotionEnvironment(branch, env); err != nil {
```

Comment:
Assuming there is a PR with the same `promotion` label, should we have a check to confirm it is ok to steal the label?

Reply:
Hum, very interesting! I think that would be feasible, especially as the exclusivity scope would typically be just that same repo. So we could list all PRs and check labels. If we find another one, we could prompt the user for confirmation, and upon positive response delete the other label.